woocommerce-services.php

```diff
@@ -1025,6 +1026,104 @@ public function reports_tabs( $reports ) {
 			return $reports;
 		}
 
+		/**
+		 * Send completed order details to Connect Server to check live rates usage.
+		 * Attached to the woocommerce_order_details_after_order_table hook
+		 *
+		 * @param $order WC_Order object containing completed order details.
+		 */
+		public function track_completed_order( $order ) {
+			// We need to recreate a temporary cart from the order
+			// so that we can create the same shipping packages from the cart
+			// to use to generate the correct cache key to return the cached
+			// shipping rates response.
+			$logger = $this->get_logger();
+
+			$temp_cart   = new WC_Cart();
```

Comment:
If we only need to determine if customer selected a WCS shipping method at checkout it is not necessary to create a temporary cart here. Or even recalculate the shipping and packages. This code should be enough:
```php
foreach( $order->get_shipping_methods() as $item_shipping ) {
					$method = WC_Shipping_Zones::get_shipping_method($item_shipping->get_instance_id());
					if ( $method instanceof  WC_Connect_Shipping_Method ) {
						// send event
					}
				};
```

Reply:
The reason why I created the temporary cart is to get the cached rates request after checkout from inside of the transient where we store it during checkout. We need to determine whether live rates associated with subscription product usage events were used; to do this, I send the live rates to the server or, if unavailable, I send the shipping packages, so that the server can repeat the request to get the rates.  However, if we can send the cached rates to the server, then there is no need to repeat the shipping rates request on the Connect Server and the request should be much quicker. 

The cache key to access the transient value for a user's checkout is computed from a hash of the shipping method's service settings and the package, to keep it unique enough to use for a customer's specific order. You can find that code [here](https://github.com/Automattic/woocommerce-services/blob/eadf71f025d53214d2628f2440f5f53d65e20f1c/classes/class-wc-connect-shipping-method.php#L439) and the relevant section is below.

```
$cache_key = sprintf(
	'wcs_rates_%s',
	md5( serialize( array( $services, $package, $custom_boxes, $predefined_boxes ) ) )
);
```

I need to recreate the shipping packages at checkout in order to create the same hash to use the same key to return the correct rates from the transient. However, since the original cart is no more, I make a new one. That's why I also needed to perform the below `ksort` on the package contents, to ensure that the package object is the same in order to create the same hash.

```
foreach ( $package['contents'] as $item_key => $item_details ) {
	ksort( $package['contents'][ $item_key ] );
}
```

Figuring this out and getting the right package formatted to create exactly the same hash for the key was probably the least fun part of all of this work, so if there is a better way to access this transient that I am unaware of, I am definitely open to any suggestions!

Reply:
I think the cart object will still exist after switching to the new hook. So you should be able to use `WC()->get _shipping_packages()` in this code here now instead of creating a new cart. If it doesn't work we could switch to using the `woocommerce_checkout_create_order` action which happens before the `woocommerce_checkout_order_processed`

Reply:
@superdav42 I've just removed the temporary cart and replaced it with the code you suggested here!